### PR TITLE
BanditBear: Add `bandit_selected_tests` option

### PR DESCRIPTION
**Pull request by sangamcse**
This adds an additional feature to run the bear
using a list of selected tests IDs.

Closes https://github.com/coala/coala-bears/issues/2386

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `corobo mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by li-boxuan** [comment]
IMO ```bandit_selected_tests``` might be a better name.

**Comment by li-boxuan** [comment]
I'm just curious about what would happen if we skip and select same test IDs.

**Comment by stellargo** [comment]
Instead of `-s` and `-t` isnt something along the lines of `--skip-test` `--perform-test` better? :p

**Comment by li-boxuan** [comment]
@stellargo ``-s`` and ``-t`` are used by bandit, not BanditBear. :stuck_out_tongue:

**Comment by stellargo** [comment]
Ahhhh sorry nvm then :)

File: bears/python/BanditBear.py [hunk 1]
```diff
@@ -24,18 +26,24 @@ class BanditBear:
...
+                          'B607'),
+                         bandit_selected_tests: typed_list(str)=()):
         """
-        :param bandit_skipped_tests:
-            The IDs of the tests ``bandit`` shall not perform. You can get
-            information about the available builtin codes at
-            https://github.com/openstack/bandit#usage.
+        :param bandit_skipped_tests: The IDs of the tests ``bandit`` shall
+                                     not perform.
+        :param bandit_selected_tests:  The IDs of the tests ``bandit`` shall
```
Comment (li-boxuan) [inline bears/python/BanditBear.py]: A redundant space here

Comment (sangamcse) [reply bears/python/BanditBear.py]: Thank you @li-boxuan. Any other review? 😛

File: bears/python/BanditBear.py [hunk 2]
```diff
@@ -47,7 +55,22 @@ def create_arguments(filename, file, config_file,
...
+        # such as selected test ID and skipped test ID are same.
+        err_pattern = re.compile(r'ERROR.*')
+        match = err_pattern.search(output[1])
+        if match:
+            warn_issue(match.group())
+
+        if not file or not output[0]:
+            return
+
+        output = json.loads(output[0])
```
Comment (virresh) [inline bears/python/BanditBear.py]: The try, catch block was removed from here ?
(Not sure though if bandit will ever give incorrect json, so just asking :sweat_smile: )

Comment (sangamcse) [reply bears/python/BanditBear.py]: try, catch block was never here, though I tried generating one incorrect json using 6/7 test files and in all the case, it either gave ERROR (which was not in json) or json output. So, IMO it is not required.

Comment (virresh) [reply bears/python/BanditBear.py]: Ah
Cool :+1:

**Comment by li-boxuan** [comment]
IMO generate_skip_decorator should be added, but seems that's not related to this issue.

File: bears/python/BanditBear.py [hunk 3]
```diff
@@ -47,7 +55,22 @@ def create_arguments(filename, file, config_file,
...
-        output = json.loads(output)
+        def warn_issue(message):
+            self.warn('While running {0}, some issues were found:'
+                      .format(self.__class__.__name__))
+            self.warn(message)
+
+        # Taking output from stderr in case bandit shows errors
+        # such as selected test ID and skipped test ID are same.
+        err_pattern = re.compile(r'ERROR.*')
+        match = err_pattern.search(output[1])
```
Comment (li-boxuan) [inline bears/python/BanditBear.py]: Can you add a line like ```stdout, stderr = output``` otherwise ```output[1]``` and ```output[0]``` are hard to understand

File: bears/python/BanditBear.py [hunk 4]
```diff
@@ -47,7 +55,22 @@ def create_arguments(filename, file, config_file,
...
+        if match:
+            warn_issue(match.group())
+
+        if not file or not output[0]:
```
Comment (li-boxuan) [inline bears/python/BanditBear.py]: I'm not sure here, but why is ```not file``` necessary?

File: bears/python/BanditBear.py [hunk 5]
```diff
@@ -7,7 +8,8 @@
 from coalib.settings.Setting import typed_list
 
 
-@linter(executable='bandit')
+@linter(executable='bandit',
+        use_stderr=True)
```
Comment (li-boxuan) [inline bears/python/BanditBear.py]: I understand the default value of ```use_stdout``` is ```True```. Not sure here, but maybe explicitly adding ```use_stdout=True``` makes the code easier to understand? (ref: http://api.coala.io/en/latest/Developers/Linter_Bears_Advanced.html)

Comment (sangamcse) [reply bears/python/BanditBear.py]: It doesn't matter as `use_stdout` is True by default. If I add it here, it will be redundant. And What's the point of adding one extra line. To understand this, one should have understanding of `linter` class and if anyone understand `linter` class, then IMO he should also know about default `use_stdout`.

File: bears/python/BanditBear.py [hunk 6]
```diff
@@ -22,19 +27,26 @@ class BanditBear:
 
     @staticmethod
     def create_arguments(filename, file, config_file,
-                         bandit_skipped_tests: typed_list(str)=
+                         bandit_skip_tests: typed_list(str)=
```
Comment (jayvdb) [inline bears/python/BanditBear.py]: need to deprecate this setting.
(and fix spaces around `=` and rebase...)

File: bears/python/ChecksSelector/ChecksSelector.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
+from coalib.settings.Setting import typed_list
```
Comment (jayvdb) [inline bears/python/ChecksSelector/ChecksSelector.py]: no; we have one module per bear.

File: bears/python/ChecksSelector/ChecksSelector.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+
+def set_checks(ignore: str = '--ignore',
```
Comment (jayvdb) [inline bears/python/ChecksSelector/ChecksSelector.py]: This method belongs in your bear, until it is migrated into coalib.

File: bears/python/BanditBear.py [hunk 9]
```diff
@@ -21,21 +27,28 @@ class BanditBear:
     CAN_DETECT = {'Security'}
 
     @staticmethod
+    @deprecate_settings(bandit_skip_tests='bandit_skipped_tests')
     def create_arguments(filename, file, config_file,
-                         bandit_skipped_tests: typed_list(str) = (
+                         bandit_skip_tests: typed_list(str) = (
```
Comment (jayvdb) [inline bears/python/BanditBear.py]: "tests" is bandit terminology.
And it is confusing terminology.

We are using `foo_ignore` and `foo_select`

File: bears/python/ChecksSelector/ChecksSelector.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+               ignore_check: typed_list(str) = (),
+               select: str = '--select',
```
Comment (jayvdb) [inline bears/python/ChecksSelector/ChecksSelector.py]: `ignore` & `select` are confusing names. `ignore_cli_arg` and `select_cli_arg` are more clear.

File: tests/python/BanditBearTest.py [hunk 11]
```diff
@@ -24,31 +26,38 @@ def load_testfile(name, splitlines=False):
...
+@generate_skip_decorator(BanditBear)
 class BanditBearTest(LocalBearTestHelper):
-    test_assert = test(
+
+    def setUp(self):
+        self.section = Section('bandit')
+        self.queue = Queue()
+        self.uut = BanditBear(self.section, self.queue)
+
+    def execute_test(testfilename, expected_results):
```
Comment (jayvdb) [inline tests/python/BanditBearTest.py]: why did you move the `test` function code here?
